### PR TITLE
[WIP] uneven input support for DDP

### LINKED ISSUE #3325: Support uneven DDP inputs with pytorch model.join
See more details: https://github.com/pytorch/pytorch/issues/38174

### PR DESCRIPTION
**Pull request by otaj**
## What does this PR do?

WIP for Uneven input support for DDP

Fixes #3325
## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/pytorch_lightning/strategies/ddp.py [hunk 1]
```diff
@@ -362,7 +367,9 @@ def reduce(
     def training_step(self, *args: Any, **kwargs: Any) -> STEP_OUTPUT:
         assert self.model is not None
         with self.precision_plugin.train_step_context():
-            return self.model(*args, **kwargs)
+            with DistCallRecorder() as recorder:
+                with Join([self.model], enable=True, dist_recorder=recorder):
```
Comment (carmocca) [inline src/pytorch_lightning/strategies/ddp.py:371]: This should be opt-in (Maybe you added this by default whilst it's WIP)

**Comment by anth2o** [comment]
Hello !
Is this PR still ongoing ?

**Comment by otaj** [comment]
> Hello ! Is this PR still ongoing ?

Hi @anth2o! Currently, there is no progress. It is somewhat still on my radar, but I've been busy with higher priority items lately.

We still want to finalize it, but it doesn't have the highest priority, honestly, so there is no ETA.

**Comment by anth2o** [comment]
Thanks for your answer, I'll try to find a solution on my end to handle this

**Comment by ssharpe42** [comment]
Is there anything on this topic being worked on currently? What is the current best suggestion in the case of using iterable datasets on separate (possibly unequal) sharded data for each gpu in training and evaluation w/ torchmetrics?

**Comment by ssharpe42** [comment]
Forgot to tag @carmocca @otaj. Also @anth2o did you find a solution here?

**Comment by ivandariojr** [comment]
This would be really useful to get consistent evaluations of datasets that may not be divisible by the number of devices. Has there been any progress on this?
